### PR TITLE
Reschedule infra failures on try jobs

### PR DESCRIPTION
**Pull request by dnfield**
Fixes https://github.com/flutter/flutter/issues/50118

Rather than failing immediately on any failure, reschedule infra failures up to 2 times.

### CODE REVIEW

**Comment by rmistry** [comment]
> /cc @rmistry

👍 👍 👍

File: app_dart/lib/src/datastore/cocoon_config.dart [hunk 1]
```diff
@@ -113,6 +113,9 @@ class Config {
 
   int get maxTaskRetries => 2;
 
+  /// The number of times to retry a LUCI job on infra failures.
+  int get luciTryJobRetries => 2;
```
Comment (CaseyHillers) [inline app_dart/lib/src/datastore/cocoon_config.dart]: nit: the variable name should mention it's for infra failure attempts, but I don't have a better name other than `maxLuciAttemptsOnInfraFailure`

File: app_dart/lib/src/request_handlers/luci_status.dart [hunk 2]
```diff
@@ -81,6 +88,78 @@ class LuciStatusHandler extends RequestHandler<Body> {
     return Body.empty;
   }
 
+  Future<void> _processCompleted({
```
Comment (CaseyHillers) [inline app_dart/lib/src/request_handlers/luci_status.dart]: Does `_taskCompleted(...)` make more sense in this situation?

File: app_dart/lib/src/request_handlers/luci_status.dart [hunk 3]
```diff
@@ -81,6 +88,78 @@ class LuciStatusHandler extends RequestHandler<Body> {
...
+    }
+    await _setCompletedStatus(
+      ref: sha,
+      builderName: builderName,
+      buildUrl: build.url,
+      result: build.result,
+    );
+  }
+
+  Future<void> _rescheduleBuild({
```
Comment (CaseyHillers) [inline app_dart/lib/src/request_handlers/luci_status.dart]: method documentation? (also for _processComplete)

File: app_dart/lib/src/request_handlers/luci_status.dart [hunk 4]
```diff
@@ -12,10 +12,12 @@ import 'package:meta/meta.dart';
 
 import '../datastore/cocoon_config.dart';
 import '../model/appengine/service_account_info.dart';
+import '../model/luci/buildbucket.dart' as bb;
```
Comment (CaseyHillers) [inline app_dart/lib/src/request_handlers/luci_status.dart]: Not a comment, but I like this prefix :)

File: app_dart/test/request_handlers/luci_status_test.dart [hunk 5]
```diff
@@ -234,6 +237,55 @@ void main() {
     );
   });
 
+  test('Reschedules an infra failure', () async {
```
Comment (CaseyHillers) [inline app_dart/test/request_handlers/luci_status_test.dart]: Does this also check that we do not get into a reschedule loop?

File: app_dart/lib/src/datastore/cocoon_config.dart (hunk 1 above)
Comment (dnfield) [reply app_dart/lib/src/datastore/cocoon_config.dart]: Going with luciTryInfraFailureRetries

File: app_dart/lib/src/request_handlers/luci_status.dart (hunk 2 above)
Comment (dnfield) [reply app_dart/lib/src/request_handlers/luci_status.dart]: Ahh this is ambiguous.  The ideas is this method should process completed jobs, it's a bad name.  I'll update and document.

File: app_dart/test/request_handlers/luci_status_test.dart (hunk 5 above)
Comment (dnfield) [reply app_dart/test/request_handlers/luci_status_test.dart]: Added another test to verify the negative side of what we do when we have exceeded the retries.

This test does verify that we increment the retries counter on the user_data.  The new test verifies that if that value >= the config value, we set the status to failure.

**Comment by dnfield** [comment]
Deploying as v364
